### ISSUE TITLE
Cover flow / Carousel view in the disk mount dialog

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#303]
I am trying to add an image view to the ADF mount dialog that mimics the cover flow / carousel style Apple once used in the Finder.
 
So far, I've found the following frameworks:

1. iCarousel
  https://github.com/nicklockwood/iCarousel

    - Pros: GPL compatible license, supports both macOS and iOS
    - Cons: Abandoned project (last commit 3 years ago), flicker bug in Cathalina
    - Conclusion: not usable in its current state (although it has great features)

2. YRCoverFlowLayout
  https://github.com/solomidSF/YRCoverFlowLayout

    - Pros: MIT licence, no bugs spotted
    - Cons: iOS only, Objective-C

3. LNICoverFlowLayout (Swift port of YRCoverFlowLayout)
  https://github.com/sumitlni/LNICoverFlowLayout

    - Pros: MIT license, Swift
    - Cons: Fails to compile the demo app on my machine

If somebody knows another suitable framework, please let me know.
Otherwise, I’ll try to port YRCoverFlowLayout from iOS to macOS.

**Comment by dirkwhoffmann** [comment]
Despite my first post, I decided to go with iCarousel 😎. Stay tuned.

**Comment by dirkwhoffmann** [comment]
To overcome the boredom of governmental imposed imprisonment, I was working on a new experimental feature: Auto screenshots (TM) 😎.

Here is how it works. If a disk is dragged into the emulator, the newly designed ADF dialog comes up:

<img width="564" alt="Bildschirmfoto 2020-03-28 um 14 41 34" src="https://user-images.githubusercontent.com/12561945/77824497-61ed3e80-7103-11ea-9207-535c215a0a16.png">

Looks a little boring, I know. To let the fun begin, first enable auto screenshots:

<img width="564" alt="Bildschirmfoto 2020-03-28 um 14 41 41" src="https://user-images.githubusercontent.com/12561945/77824510-78939580-7103-11ea-93a2-dccbeaf3a10a.png">

To generate screenshots, hit df0: and do what you would usually do with an Amiga. In case of Shadow of the Beast, just sit in front of your computer and wait until it has finished loading (which literally takes forever). 


From now on, when you insert the disk again, the dialog looks like this 😃: 

<img width="564" alt="Bildschirmfoto 2020-03-28 um 14 40 39" src="https://user-images.githubusercontent.com/12561945/77824543-c27c7b80-7103-11ea-8e48-c87d2cc34f63.png">

I have to admit that this is kind of a marketing picture 🙄. Because the emulator takes screenshots based on a hard-coded schedule, it sometimes grabs an empty screen 🤭.

<img width="564" alt="Bildschirmfoto 2020-03-28 um 14 41 08" src="https://user-images.githubusercontent.com/12561945/77824557-dd4ef000-7103-11ea-8d97-e7dea1ef2c9d.png">

All images are stored locally in the Application Support folder:

<img width="882" alt="Bildschirmfoto 2020-03-28 um 14 59 39" src="https://user-images.githubusercontent.com/12561945/77824695-d2e12600-7104-11ea-9f97-5e9e66cf3a38.png">

The directory name is the fnv64 fingerprint of the inserted disk. 

Right now, up to 6 screenshots are taken. The intervals are hard-coded inside the following function. Maybe we can come up with better timing scheme. One could also allow an infinite number screenshots with a logarithmic time scale (e.g., by doubling the interval between screenshots every time a new one is taken).

```
void
Amiga::serviceScrEvent()
{
    static int delay[] = { 20, 30, 30, 30, 30, 30, 0 };
    
    int num = numAutoScreenshots();
    takeAutoScreenshot();
    
    putMessage(MSG_AUTOSCREENSHOT, agnus.slot[SCR_SLOT].data);
    
    if (delay[num]) {
        agnus.rescheduleRel<SCR_SLOT>(SEC(delay[num]));
    } else {
        agnus.cancel<SCR_SLOT>();
    }
```

**Comment by dirkwhoffmann** [comment]
OK, let's bring this to the next level. There are already two storages for snapshots: Auto-taken snapshots and user-taken snapshots. And there is already a dialog for managing them: 

<img width="588" alt="Bildschirmfoto 2020-03-28 um 17 24 38" src="https://user-images.githubusercontent.com/12561945/77828049-a4ba1100-7119-11ea-86ca-7d0bc3626a23.png">

I think the straight-forward thing to do now is to add two similar storages for screenshots: Auto-taken screenshots and user-taken screenshots. User-taken screenshots could be generated by pushing a button in the toolbar just like user-snapshots are generated. In addition, there could be a screenshot dialog similar to the shown above. Now, when a disk is dragged in, the emulator first searches for user-saved screenshots. If it founds some, it displays them in the iCarousel view. If it doesn't find any, it'll display the auto-taken screenshots instead.

**Comment by mithrendal** [comment]
Looks good. 🤤 Maybe we could display  snapshots  here instead of pictures only ? Snapshots also have picture plus a gamestate. On selection you could beam yourself into the game state. The storage place of the gamestates could be the same as you did for the pictures. Hm I don’t know 🤔 game state is much bigger who does the housekeeping, how to delete a snapshot ? 
🙃Good question how to delete individual pictures in your  current concept?

**Comment by dirkwhoffmann** [comment]
> you never reach the end of the list because the content randomly repeats

My Netflix list has become infinite, but never repeats. Don't know how I did it 🙄. Now I am struggling with the Hilbert hotel paradox. You can kick as many customers out of the hotel as you want without creating vacancy (OK, usually it's more positively formulated: There's always room for a new client). 

> A game state browser?

Unfortunately, I think we have to backpedal a bit, because things are tricky.
- My experience with VirtualC64 has shown that the snapshot format is likely to change from release to release. This means that we shouldn't pursue any approach that is based on storing snapshots persistently in a data base. Originally, I wanted to have snapshots to ease game play. Like saving a snapshot before making a hard decision in a game: Should I go through the left door or the right door? Should I rescue the brunette princess or the blonde? 

- Another issue is this: Screenshots are bound to the fingerprint of the most recently inserted disk. Now, many games have multiple disks. This means that most screenshots are bound to the second or third disk of the game. This is OK for screenshots, but it wouldn't work for snapshots.

- I think I have to dump my previous screenshot idea, too. The problem really is size. For VirtualC64, a lot of users asked my for screenshot support in the past. Hence, I think this is really something people want to have. However, they want to have high-quality screenshots and not just thumbnail images. Now, the problem is size: If we saved high-quality screenshots instead of thumbnails, we cannot just dump them into Application Support and never delete them. 

My new idea is to add kind of "Like" button that makes auto-taken screenshots persistent. All other screenshots will be overwritten each time you reuse a disk. I think it's important to keep it simple, because it's still just an eye-candy feature.

**Comment by dirkwhoffmann** [comment]
I’ve reworked the entire thing on the lessons learned. Here is my new approach:

When the user inserts a yet unknown disk, the dialog looks as follows:

<img width="563" alt="Bildschirmfoto 2020-03-30 um 14 14 02" src="https://user-images.githubusercontent.com/12561945/77912331-a824d800-7292-11ea-858b-10b6a1cf4422.png">

Now imagine that the user hits df0: and plays a couple of minutes:

imagine ... imagine ... imagine 🥱

The next time, he inserts the disk, it will look like this:

<img width="563" alt="Bildschirmfoto 2020-03-30 um 14 05 49" src="https://user-images.githubusercontent.com/12561945/77912358-b541c700-7292-11ea-9499-303f4c50b783.png">

The dialog automatically displays all screenshots that were taken in the background while the user was playing.

Now, by clicking the star below the center image, the user can „like“ the image which moves it to the 'favorites'. So, let’s do this for a couple of nice images:

click .. click ... click 🤤

Now, when the disk is inserted again, the emulator finds the favorites on disk and displays them instead of the auto-captured images. In my example, this looks like this: 😎 

<img width="563" alt="options" src="https://user-images.githubusercontent.com/12561945/77912433-d60a1c80-7292-11ea-9e8c-3b0e5b283329.png">

There are some more options. By using the arrow buttons, the user is able to shuffle the images around (in case he doesn’t like the ordering). And by clicking on the path folder, the finder comes up. Going one level up in the finder reveals the directory structure. Auto-screenshots are kept in folder 'auto' and the selected favorites are kept in folder 'user':

<img width="547" alt="Bildschirmfoto 2020-03-30 um 14 25 56" src="https://user-images.githubusercontent.com/12561945/77912458-e5896580-7292-11ea-9a7d-2ab8a89c0fcb.png">

All auto-taken screenshots are deleted automatically when a new disk is inserted. All favorites are kept until the user deletes them. Deleting is easy. When the favorites are displayed, the star icon changes to a trashcan icon.

**Comment by mithrendal** [comment]
Ok lets execute the concept ...

>The next time, he inserts the disk, it will look like this ... Now, by clicking the star below the center image, the user can „like“ the image which moves it to the 'favorites'.

Ok good ... reinserting the disk lets me select my favourites  from the auto-taken screenshots 😋 ... very nice I like this one  
![grafik](https://user-images.githubusercontent.com/17108995/77927328-66eaf300-72a7-11ea-890d-706f59b397ff.png)

 

>All auto-taken screenshots are deleted automatically when a new disk is inserted. 

Oh no. They are deleted before I can select one.😭

**Comment by dirkwhoffmann** [comment]
>  did I something wrong ?

You are not eligible to post today, because you said you are two days off from your computer.
Hmm, OK, maybe you're using your smartphone 🤔.

> They are deleted before I can select one.

No, inserting the disk happens when you close the ADF dialog (e.g., by pressing Df0:).  Hence, they will be there when you use the disk a second time. But this also means that you only have one chance to add the cool pics to the favorites 😎...

**Comment by dirkwhoffmann** [comment]
My favorites 😎

<img width="569" alt="Bildschirmfoto 2020-03-30 um 17 14 02" src="https://user-images.githubusercontent.com/12561945/77929391-e7aaee80-72a9-11ea-945c-f18b0e6afdfd.png">

**Comment by mithrendal** [comment]
>My favorites 😎
I love it ! 

>You are not eligible to post today, because you said you are two days off from your computer.
Gotcha. 😁 Ok how o argue now.🤔 ... I meant duration of calendar days, which implies that everything between 24h and 48h is two days.  I was off 24h and 1 minute which is 2 calendar days, no ?

**Comment by dirkwhoffmann** [comment]
>  I was off 24h and 1 minute which is 2 calendar days

🤔 Hmm, I need to recalculate that in a quiet moment.

I was surprised you left the house anyway. Didn't you look out of the window? 😬

![zom](https://user-images.githubusercontent.com/12561945/77936592-648e9600-72b3-11ea-8e29-7fe186607d9c.png)

**Comment by dirkwhoffmann** [comment]
I think the new approach is quite handy. It's really simple to click some nice artwork together 😎:

<img width="566" alt="Bildschirmfoto 2020-03-30 um 18 36 23" src="https://user-images.githubusercontent.com/12561945/77938054-77a26580-72b5-11ea-9dff-7ae599444831.png">

<img width="566" alt="Bildschirmfoto 2020-03-30 um 18 37 36" src="https://user-images.githubusercontent.com/12561945/77938115-8ee15300-72b5-11ea-8eee-29ce479b18c4.png">

With the new approach, I might even be able to reactivate my idea about a screenshot browser. Since auto-screenshots get deleted automatically (at least when using a disk twice) and favorites can be deleted manually, disk capacity doesn't seem to be a major issue any more 🤔.

**Comment by mithrendal** [comment]
Yes yes yes 🤤... With an option for netflix infinity list style 😜

**Comment by dirkwhoffmann** [comment]
Here is the new (experimental) snapshot / screenshot browser: 

<img width="660" alt="Bildschirmfoto 2020-03-31 um 18 47 04" src="https://user-images.githubusercontent.com/12561945/78053031-48a4f600-7380-11ea-995e-0d167cb66185.png">

What I don't like about the current concept is that snapshots and screenshots are treated so differently. Firstly, snapshots are volatile, but screenshots are persistent (at least the favorites). I think that's acceptable due to the reasons explained in one of my other posts. Secondly, snapshots are organized in a flat hierarchy (in two lists, one for auto-taken snapshots and one for user-taken snapshots) whereas screenshots are organized hierarchically (they are bound to a specific ADF to make the artwork stuff work). This is what I don't like. I think many users wouldn't understand the conceptual difference. Maybe it makes sense to organize snapshots hierarchically, too. But to make things transparent the hierarchy would have to be made visible (in some way) inside the browser (kind of a directory list, and the thumbnail pictures would be the directory items).

**Comment by mithrendal** [comment]
> ... to make things transparent the hierarchy would have to be made visible (in some way) inside the browser (kind of a directory list, and the thumbnail pictures would be the directory items)


Maybe for the game states we better need  ... oh I already said this above 🤐

Maybe for the ~~game states~~  screenshots  we better need something different, more organised. A ~~game states~~ screenshot browser?

Menu-> open ~~game states~~ screenshot browser

Which opens an extra window. With a list. Each row a different game for which the user explicitly saved a ~~game state~~ screnshot. Each column a picture of a user saved ~~game states~~ screenshot for the game of that row. Can scroll vertical for the different games or horizontal for the game states of particular game.... like netflix, no ? 😭😭😭 dirk won't give me my netflix drug replacement thing ...

**Comment by dirkwhoffmann** [comment]
> Which opens an extra window. With a list. Each row a different game ...

I get your point. Spinning the idea further leads to a powerful emulator frontend that not only supports ADF files, but C64 files as well. It would be a Netflix like UI with options such as „Launch in VirtualC64“ and „Launch in vAmiga“. I think FSUAE went that way to some extend with the Arcade frontend. However, I don’t want to start such a big project now (at least not before I’ve written an Atari 2600 emulator 🤤).

Right now, I am trying to clean up the coding mess on the dev branch by bringing the already implemented stuff into some useful state. My current plan is to move the already existing screenshot management tools from the ADF dialog into the combined snapshot / screenshot dialog. The splitted table view will go away. It will be replaced by something like this:

<img width="908" alt="Bildschirmfoto 2020-04-01 um 18 37 21" src="https://user-images.githubusercontent.com/12561945/78164189-ab15f900-7449-11ea-99da-71447e1f1524.png">

The ADF dialog will be simplified a bit. When a disk is inserted, it first searches for screenshots in the favorites folder. If images can be found, those images are displayed as eye candy. If there are no favorite images, the auto-saved screenshots are displayed. If there are no images at all, a default image (the one with the white camera) is displayed. 

For some reason, it seems to take me forever to code this 😖. Thus I am getting a little bit frustrated about myself, because I am wasting so much time on this features. On the other hand, isn’t this what we are supposed to do? Sitting at home, wasting time, and dreaming about the day when toilet paper will be back in store shelves 😃?

**Comment by mithrendal** [comment]
>Thus I am getting a little bit frustrated about myself, because I am wasting so much time on this feature

don't ask me for the time spent with the fullscreen thing in vc64web 🙄.  But hey, we should only do what we like to do or what we like to think ... vive l'inspiration libéré. But I get your point, it is like being a great pirate captain with a big destroyer that hunts a small shallop with a small trunk of gold coins ... and the shallop is quick and agil ... 🤔 ... it won't let you enter so easily ... 😫... it is so agil ... and oops another of that quick turn maneuver ... 😀

**Comment by dirkwhoffmann** [comment]
> don't ask me for the time spent with the fullscreen thing in vc64web

Yeaahhhh. giving up is not an option! There are more rounds to come. Stay tuned...

![rocky](https://user-images.githubusercontent.com/12561945/78188716-fc85ae80-7470-11ea-81d1-55a6e73415be.jpg)

Need to get back to Netflix... Rambo II is playing ....

**Comment by dirkwhoffmann** [comment]
The sky is clearing up. I think I have a pretty neat solution now. It works as follows: 

When a disk is inserted the first time, the ADF dialog appears like this:

<img width="601" alt="Bildschirmfoto 2020-04-02 um 17 39 40" src="https://user-images.githubusercontent.com/12561945/78269626-22ab5d00-750a-11ea-8546-8aee65835cd5.png">

Simple and clean. When a known disk is inserted, the dialog provides some additional eye candy: 

<img width="596" alt="Bildschirmfoto 2020-04-02 um 17 00 13" src="https://user-images.githubusercontent.com/12561945/78269676-35259680-750a-11ea-9f18-b0e1cfcf3104.png">

The auto-saved / user-saved selector is history. If there are user-saved images (aka favorites), those images are display. If not, the auto-taken images are displayed.

Screenshot management happens inside the new snapshot / screenshot browser. While watching Rambo on Netflix yesterday 🦾, I discovered that iCarousel has a freaking cool Time Machine layout mode 🤤. Now, the user can browse through all snapshots and screenshots in "Time Machine mode" and do the same things that were possible in the ADF dialog before (moving auto-taken screenshots to the favorites, deleting screenshots, changing the arrangement order). 

<img width="745" alt="Bildschirmfoto 2020-04-02 um 17 39 26" src="https://user-images.githubusercontent.com/12561945/78269886-71f18d80-750a-11ea-8c5b-129d489177d2.png">

There are still some issues left (i.e., when browsing through snapshots. I didn't really test the snapshot view yet which means that it is definitely broken by now 🙄).

**Comment by dirkwhoffmann** [comment]
Just encountered another issue: 

Let's assume, two emulator instances are running in parallel with the same disk in df0 😳 (the disk in df0 decides where the screenshots are stored). Both emulator instances would record screenshots and push them back into the Application Support folder from time to time. This creates chaos, because they would constantly overwrite the images of the other instance.

I think I have to jeopardize the idea of saving auto-taken screenshots on disk. Only user-taken screenshots should be saved. This means that the ADF dialog won't show auto-taken screenshots if there are no user-taken screenshots. That's OK, I think, because I already changed the dialog to only show user-taken screenshots. This is because auto-taken are usually not that visually appealing (multiple copies of the same image, empty screens etc.)

**Comment by dirkwhoffmann** [comment]
Oops, I did it again 🙄. I've reworked the whole thing 🥵. Over night, I've come to the conclusion that snapshots and screenshots do need a different handling, because they are simply not the same. E.g., snapshots have a natural temporal ordering which means that a time-machine like view is adequate. Screenshots don't have this temporal character. For them, any kind of gallery view is the natural choice. To make a long story short: The shared snapshot / screenshot browser is history (it had a short life-span anyway). 

In the next release, there will be two toolbar sections. One for snapshots and one for screenshots: 

<img width="787" alt="Bildschirmfoto 2020-04-06 um 17 43 19" src="https://user-images.githubusercontent.com/12561945/78579174-cfe6e380-7830-11ea-9ed9-118abf02bffd.png">

Clicking the time-machine icon in the snapshot section opens the snapshot browser. The browser works very similar to Apple’s Time Machine. You can move back and forth along the z-axis with a nice animation. Snapshots can be saved, restored, or deleted:

<img width="442" alt="Bildschirmfoto 2020-04-06 um 18 08 34" src="https://user-images.githubusercontent.com/12561945/78579732-aaa6a500-7831-11ea-88cb-347476298df7.png">

The split view is gone, because I wanted to simplify things.

The folder icon in the screenshot section opens up the screenshot browser. It pretty much resembles what was there in the ADF dialog before. In the manually-taken-screenshot view, the user can delete screenshots or reorder them.  In the automatically-taken-screenshot view, there is a star button to move an auto-taken screenshot to the user-taken screenshot (which makes it persistent).

<img width="588" alt="Bildschirmfoto 2020-04-06 um 17 54 21" src="https://user-images.githubusercontent.com/12561945/78579788-bbefb180-7831-11ea-8ce2-daf446315790.png">

The ADF dialog has been simplified, too. Because screenshots are now managed inside the new screenshot browser, all buttons are gone. The dialog simply displays the manually saved screenshots (if any).

<img width="606" alt="Bildschirmfoto 2020-04-06 um 17 54 00" src="https://user-images.githubusercontent.com/12561945/78579816-cb6efa80-7831-11ea-8cc0-c4cf74528289.png">

There is another change. Screenshots are now bound to the first disk inserted into df0. I think this resembles the typical use case. Inserting some game boot disk, hitting reset, playing the game. 

I hope that I still like that solution tomorrow. If yes, I can switch back to low-level stuff. There is some more CPU work to do 😎.

**Comment by mithrendal** [comment]
Inspiration is key ! The driver of true artists greatest work. 😎👍🏻

**Comment by dirkwhoffmann** [comment]
I still love it 😃. Hence, the issue can be closed, I think.
